iquip/apps/builder.py

```diff
@@ -0,0 +1,419 @@
+    """
+
+    def __init__(self, name: str, parent: Optional[QWidget] = None):
+        """Extended.
+        
+        Args:
+            name: See the attributes section in _BaseEntry.
+        """
+        super().__init__(parent=parent)
+        self.name = name
+        # widgets
+        self.nameLabel = QLabel(name, self)
+        # layout
+        self.layout = QHBoxLayout(self)
+        self.layout.addWidget(self.nameLabel)
+        self.setLayout(self.layout)
+
+    def value(self) -> Any:
+        """Returns the entered or selected value.
+        
+        This must be overridden in the subclass.
+        """
+        raise NotImplementedError
+
+
+class _BooleanEntry(_BaseEntry):
+    """Entry class for a boolean value."""
+
+    def __init__(self, name: str, default: bool = False, parent: Optional[QWidget] = None):
+        """Extended.
+        
+        Args:
+            default: The default value. If it does not exist, it is set to False.
+        """
+        super().__init__(name, parent=parent)
+        # widgets
+        self.checkBox = QCheckBox(self)
+        self.checkBox.setChecked(default)
+        # layout
+        self.layout.addWidget(self.checkBox)
+
+    def value(self) -> bool:
+        """Overridden.
+        
+        Returns the status of the checkBox.
+        """
+        return self.checkBox.isChecked()
+
+
+class _EnumerationEntry(_BaseEntry):
+    """Entry class for an enumeration value."""
+
+    def __init__(
+        self,
+        name: str,
+        choices: List[str],
+        default: Optional[str] = None,
+        parent: Optional[QWidget] = None
+    ):
+        """Extended.
+        
+        Args:
+            choices: The pre-defined candidates.
+            default: The default value. If it does not exist, it is set to the first candidate.
+        """
+        super().__init__(name, parent=parent)
+        # widgets
+        self.comboBox = QComboBox(self)
+        for choice in choices:
+            self.comboBox.addItem(choice)
+        if choices:
+            self.comboBox.setCurrentText(choices[0] if default is None else default)
+        # layout
+        self.layout.addWidget(self.comboBox)
+
+    def value(self) -> str:
+        """Overridden.
+        
+        Returns the value of the comboBox.
+        """
+        return self.comboBox.currentText()
+
+
+class _NumberEntry(_BaseEntry):
+    """Entry class for a number value.
+    
+    Attributes:
+        scale: The scale factor that actually applies.
+        type: The type of the value. If "int", value() returns an integer value.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        unit: str,
+        scale: float,
+        step: float,
+        min: Optional[float],  # pylint: disable=redefined-builtin
+        max: Optional[float],  # pylint: disable=redefined-builtin
+        ndecimals: int,
+        type: str,  # pylint: disable=redefined-builtin
```

Comment:
If you take the type like this (by `str`), the input domain is too broad, because it only takes `"int"` or `"float"`.
I would rather use a boolean type parameter, e.g., `isInt: bool`.

Reply:
As I said above, the arguments of the constructor would be set automatically, not explicitly.
Concretly, the response of the proxy server which is the return value of `artiq.master.experiments.ExperimentDB.examine()` is passed to the arguments.
So currently we cannot modify the names..